Add getting-started to part 2 (#21)
ISSOtm/gb-asm-tutorial: src/part2/getting-started.md

# Getting started

> In this lesson we will start a new project from scratch, an arkanoid clone called "Unbricked"! (Or any other name you like, as this is *your* project)

Reviewer: [Breakout](https://en.wikipedia.org/wiki/Breakout_%28video_game%29) ([Arkanoid](https://en.wikipedia.org/wiki/Arkanoid)) (I personally recognize the name "Breakout" more, and it was first)
```suggestion
In this lesson we will start a new project from scratch, a [Breakout](https://en.wikipedia.org/wiki/Breakout_%28video_game%29) ([Arkanoid](https://en.wikipedia.org/wiki/Arkanoid)) clone called "Unbricked"! (Or any other name you like, as this is *your* project)
```

> Open a terminal and make new directory, like you did for "Hello, world!" (`mkdir unbricked`), and then enter it (`cd unbricked`).

Reviewer: ```suggestion
Open a terminal and make a new directory (`mkdir unbricked`), and then enter it (`cd unbricked`), just like you did for "Hello, world!".
```

> Start by creating a file called `main.asm`, and include hardware.inc in your code.

Reviewer: Should all filenames be monospace?
```suggestion
Start by creating a file called `main.asm`, and include `hardware.inc` in your code.
```

```
line 1
```

Then create a header. Remember from part 1 that the header includes some information the Game Boy relies on, so you don't wanna accidentally leave it out.

```
lines 3-7
```

The header jumps to `EntryPoint`, so let's write that now:

```
lines 9-18
```

> The next few lines wait until "VBlank". This basically means that the Game Boy's PPU has went to sleep for a short time, during which we can do things like turning off the screen.

Reviewer 2: `has went to sleep` doesn't sound right here

> The next few lines wait until "VBlank". This basically means that the Game Boy's PPU has went to sleep for a short time, during which we can do things like turning off the screen.

Reviewer: If you haven't already, would it be a good idea to start with a high-level explanation of what VBlank and HBlank even are? Just summarizing the idea of a beam going top to bottom, left to right, with some overscan in both directions during which the screen isn't being drawn on and things can happen like you modifying VRAM yourself. Mentioning that the origin is from the literal beam inside a CRT TV so it doesn't seem arbitrary.

If I saw a paragraph+diagram explaining that it would give a lot more context for further discussion of what the PPU and CPU are doing, and when VBlank is relative to everything else going on, and why they have these time limits (HBlank is short because the beam should quickly move on to the next row, etc).

Author: I wanna know what ISSO's plans are for that stuff, but I'm open to explaining it here.

Author: Judging by Lesson 2, it looks like VBlank is a topic for even later.
> OAM accessibility will be glossed over, as explaining the concept of VBlank is a bit of a rabbit hole, and doing the update just after the "pacing wait" will mask that problem.

Author: Also this:
https://github.com/ISSOtm/gb-asm-tutorial/pull/21/files#r918407321

> Turning off the screen is important because it puts the PPU to sleep until the screen is turned back on. You can only load new graphics while the PPU is sleeping, so this gives us time to load our tiles.

Reviewer 2: This is a bit confusing. Previously we say that during "VBlank the PPU goes to sleep", but then we affirm that we need to turn the screen off during VBlank to put the PPU to sleep. Maybe we could clarify?

Speaking of tiles, we're going to load some into VRAM next, using the following code:
```
lines 20-31
```

> This loop might look familiar if you've read part 1. It copies `Tiles` to `$9000`, which is a part of VRAM where tiles are stored. To get the length of `Tiles`, we use another label at the end, called `TilesEnd`, and get the difference!

Reviewer 2: It's `a part` or `the part` of the VRAM where tiles are stored?

Note that our code still doesn't have anything called `Tiles`. We'll get to that later!

We're almost done now. Next, write another loop, this time for copying the tilemap, which will organize our tiles on the screen.

```
lines 33-44
```

You might notice that while this code is exactly the same, the 3 values loaded into `de`, `hl`, and `bc` are different. These determine the source, destination, and size of the copy, respectively.

Finally, we are going to turn the screen back on and initialize a background palette. hardware.inc provides the constants `LCDCF_ON` and `LCDCF_BGON`, which turn on the screen and the background respectively. We can "combine" these together using `|` (the or operator).

```
lines 46-55
```

There's one last thing we need before we can build the rom, and that's our graphics. We're going to be drawing the following screen:

![Layout of unbricked](https://github.com/ISSOtm/gb-asm-tutorial-part2/blob/main/tilemap.png?raw=true)

> In the "Hello, world!" lesson, you saw graphics which were wirtten out by hand in hexadecimal. This time, we're going to write our graphics in a more friendly way; by assigning a character to each shade! To do this, type `dw`, followed by a space, a backtick (\`) and a series of 8 characters; by default these are 0, 1, 2, and 3, from lightest to darkest. This defines a row of 8 pixels.

Reviewer 2: ```suggestion
In the "Hello, world!" lesson, you saw graphics which were written out by hand in hexadecimal. This time, we're going to write our graphics in a more friendly way; by assigning a character to each shade! To do this, type `dw`, followed by a space, a backtick (\`) and a series of 8 characters; by default these are 0, 1, 2, and 3, from lightest to darkest. This defines a row of 8 pixels.
```

Open a terminal and make a new directory (`mkdir unbricked`), and then enter it (`cd unbricked`), just like you did for "Hello, world!".

Start by creating a file called `main.asm`, and include `hardware.inc` in your code.

```
line 1
```

Then create a header. Remember from part 1 that the header includes some information the Game Boy relies on, so you don't wanna accidentally leave it out.

```
lines 3-7
```

The header jumps to `EntryPoint`, so let's write that now:

```
lines 9-18
```

The next few lines wait until "VBlank".
VBlank basically means that the Game Boy's PPU has gone to sleep for a short time, during which we can do things like turning off the screen. 

Turning off the screen is important because the PPU may wake back up at while we're still working.
By turning the screen off completely, we ensure that our code won't be interrupted.
You can only load new graphics while the PPU is sleeping, so this gives us as much time as we need to load our tiles.

Speaking of tiles, we're going to load some into VRAM next, using the following code:
```
lines 20-31
```

This loop might look familiar if you've read part 1.

> It copies `Tiles` to `$9000`, which is the part of VRAM where tiles are stored.

Author: > which is a/the part of VRAM where tiles are stored.
This originally said "a" but @avivace suggested I change it, maybe he wants to give some input on this?

> It copies `Tiles` to `$9000`, which is the part of VRAM where tiles are stored.

Reviewer 3: $9000 is not the beginning of that area, so I think calling it "the" is wrong.

> It copies `Tiles` to `$9000`, which is the part of VRAM where tiles are stored.

Reviewer 2: My suggestion meant to try to clarify what $9000 was (a choice in a range?)

> It copies `Tiles` to `$9000`, which is the part of VRAM where tiles are stored.

Reviewer: `_VRAM + $1000`?

To get the length of `Tiles` we use another label at the end, called `TilesEnd`, and subtract the two to find the difference.

Note that our code still doesn't have a lable called `Tiles`.
We'll get to that later!

We're almost done now.
Next, write another loop, this time for copying the tilemap, which will organize our tiles on the screen.

```
lines 33-44
```

You might notice that while this code is exactly the same, the 3 values loaded into `de`, `hl`, and `bc` are different.
These determine the source, destination, and size of the copy, respectively.

Finally, we are going to turn the screen back on and initialize a background palette.
hardware.inc provides the constants `LCDCF_ON` and `LCDCF_BGON`, which enable the screen and the background.
We must "combine" these together using `|` (the or operator).

```
lines 46-55
```

There's one last thing we need before we can build the rom, and that's our graphics.
We're going to be drawing the following screen:

> ![Layout of unbricked](https://github.com/ISSOtm/gb-asm-tutorial-part2/blob/main/tilemap.png?raw=true)

Author: Yes, I agree. I can add `assets/part2` in this PR and we can move part 1 into `assets/part1` in another.

In the "Hello, world!" lesson, you saw graphics which were written out by hand in hexadecimal.
This time, we're going to write our graphics in a more friendly way; by assigning a character to each shade.

> To do this, type `dw`, followed by a space, a backtick (\`) and a series of 8 characters; by default these are 0, 1, 2, and 3, going from lightest to darkest.
> This defines a row of 8 pixels.

Reviewer 3: ```suggestion
We will use `dw` instead of `db` (the difference between these two will be explained later); and for each row of pixels, instead of writing [the bitplanes](../part1/tiles#encoding) as raw numbers, we will use a backtick (\`) followed by 8 characters.
Each character defines a single pixel, intuitively from left to right; it must be one of 0, 1, 2, and 3, representing the corresponding color index in [the palette](../part1/palettes).
```

> ```rgbasm,linenos
> ; For example:
> Tiles:

Reviewer 3: ```suggestion
For example:

```rgbasm
```

	dw `01230123
```

> (A note about OPT g could be added here?)

Reviewer 3: This would be a footnote attached to the "0, 1, 2, 3" snippet. It must also include [a link to the docs](https://rgbds.gbdev.io/docs/v0.5.2/rgbasm.5/#Changing_options_while_assembling).

> We already have tiles made for this project, so you can copy [this premade file](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tileset.asm), and paste it at the end of your code.

Reviewer 3: The lines must be included inline (possibly with a `max-height` applied to the block), especially since there is a handy "copy to clipboard" button.

Reviewer 2: I don't think #17 should be blocking, as it involves a large amount of work to set everything up. Please just start with the example code "imported"..

> Then copy the tilemap from [this file](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tilemap.asm), and paste it after the `TilesEnd` label.

Reviewer 3: Same.

> You can try building the rom at this point, using the following commands in your terminal:

Reviewer 3: ```suggestion
You can build the ROM now, by running the following commands in your terminal:
```

```console
$ rgbasm -L -o main.o main.asm
$ rgblink -o unbricked.gb main.o
$ rgbfix -v -p 0xFF unbricked.gb
```

If you run this in your emulator, you should see the following:

(Screenshot pending :P)

That white square seems to be missing!

> If you paid attention to your tiles earlier, you may have noticed this comment:

Reviewer 3: ```suggestion
You may have noticed this comment earlier, somewhere in the tile data:
```

```
lines 135-140
```

> The logo tiles were left intentionally blank so that you could customize this project.

Reviewer 3: ```suggestion
The logo tiles were left intentionally blank so that you can choose your own.
```

> If you feel up to it, you can try creating your own logo by hand, or you can copy one of the following logos into your code:

Reviewer 3: ```suggestion
You can use one of the following pre-made logos, or try coming up with your own!
```

## RGBDS Logo
[Source](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/rgbds.asm)
![The RGBDS Logo](https://github.com/ISSOtm/gb-asm-tutorial-part2/blob/main/rgbds.png?raw=true)

## Duck
[Source](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/duck.asm)
![A pixel-art duck](https://github.com/ISSOtm/gb-asm-tutorial-part2/blob/main/duck.png?raw=true)

## Tail
[Source](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tail.asm)
![A silhouette of a tail](https://github.com/ISSOtm/gb-asm-tutorial-part2/blob/main/tail.png?raw=true)

> Build your game again and your logo of choice should appear in the bottom right!

Reviewer 3: ```suggestion
Replace the blank tiles with the new graphics, build the game again, and you should see your logo of choice in the bottom-right!
```

# Getting started

In this lesson, we will start a new project from scratch.

> We will make a [Breakout](https://en.wikipedia.org/wiki/Breakout_%28video_game%29) / [Arkanoid](https://en.wikipedia.org/wiki/Arkanoid) clone, which we'll "Unbricked"!

Reviewer 2: ```suggestion
We will make a [Breakout](https://en.wikipedia.org/wiki/Breakout_%28video_game%29) / [Arkanoid](https://en.wikipedia.org/wiki/Arkanoid) clone, which we'll call "Unbricked"!
```

(Though you are free to give it any other name you like, as it will be *your* project)

Open a terminal and make a new directory (`mkdir unbricked`), and then enter it (`cd unbricked`), just like you did for ["Hello, world!"](../part1/hello-world).

Start by creating a file called `main.asm`, and include `hardware.inc` in your code.

```
line 1
```

`hardware.inc` is a file that provides constants which allow you to interface with the rest of the Game Boy.
When you write code, your instructions are only read by the CPU.

> To access other parts of the system, like the screen, buttons, or audio, you use special registers known as "I/O" registers.

Reviewer 2: Maybe a link to https://en.wikipedia.org/wiki/Memory-mapped_I/O could help here?

Each character defines a single pixel, intuitively from left to right; it must be one of 0, 1, 2, and 3, representing the corresponding color index in [the palette](../part1/palettes).

::: tip

0, 1, 2, and 3 aren't the only options for writing graphics.
You can use [`OPT g`](https://rgbds.gbdev.io/docs/v0.5.2/rgbasm.5/#Changing_options_while_assembling) to modify these characters to your liking.

:::

For example:

```rgbasm
	dw `01230123 ; This is equivalent to `db $55,$33`
```

We already have tiles made for this project, so you can copy [this premade file](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tileset.asm), and paste it at the end of your code.

Then copy the tilemap from [this file](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tilemap.asm), and paste it after the `TilesEnd` label.

You can build the ROM now, by running the following commands in your terminal:

```console
$ rgbasm -L -o main.o main.asm
$ rgblink -o unbricked.gb main.o
$ rgbfix -v -p 0xFF unbricked.gb
```

If you run this in your emulator, you should see the following:

> (Screenshot pending :P)

Reviewer 2: TODO

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:header}}
{{#include ../../unbricked/getting-started/main.asm:header}}
```

The header jumps to `EntryPoint`, so let's write that now:

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:entry}}
{{#include ../../unbricked/getting-started/main.asm:entry}}
```

The next few lines wait until "VBlank", which is the only time you can safely turn off the screen (doing so at the wrong time could damage a real Game Boy, so this is very crucial). We'll talk more about VBlank later.

Turning off the screen is important because loading new tiles while the screen is on is tricky—we'll touch on how to do that in Part 3.

Speaking of tiles, we're going to load some into VRAM next, using the following code:

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:copy_tiles}}
{{#include ../../unbricked/getting-started/main.asm:copy_tiles}}
```

This loop might be [reminiscent of part Ⅰ](../part1/jumps#conditional-jumps).

> It copies `Tiles` to `$9000`, which is the part of VRAM where our tiles are going to be stored.

Reviewer 3: ```suggestion
It copies starting at `Tiles` to `$9000` onwards, which is the part of VRAM where our [tiles](../part1/tiles.md) are going to be stored.
```

> `$9000` is the first background tile, so it's assigned an ID of 0, and every tile after it is just one ID higher.

Reviewer 3: Isn't this redundant with Part I?

Author: Avivace [suggested I clarify it](https://github.com/ISSOtm/gb-asm-tutorial/pull/21#discussion_r922537838).

Reviewer 3: ```suggestion
Recall that `$9000` is where the data of background tile $00 lies, and the data of subsequent tiles follows right after.
```

Rather than writing the non-descript number `%10000001` (or $81 or 129, to taste), we make use of two constants graciously provided by `hardware.inc`: `LCDCF_ON` and `LCDCF_BGON`.
When written to [`rLCDC`](https://gbdev.io/pandocs/LCDC), the former causes the PPU and screen to turn back on, and the latter enables the background to be drawn.
(There are other elements that could be drawn, but we are not enabling them yet.)
Combining these constants must be done using `|`, the *binary "or"* operator; we'll see why later.

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:end}}
{{#include ../../unbricked/getting-started/main.asm:end}}
```

There's one last thing we need before we can build the ROM, and that's the graphics.
We will draw the following screen:

![Layout of unbricked](../assets/part2/img/tilemap.png)

In `hello-world.asm`, tile data had been written out by hand in hexadecimal; this was to let you see how the sausage is made at the lowest level, but *boy* is it impractical to write!
This time, we will employ a more friendly way, which will let us write each row of pixels more easily.

> We will use `dw` instead of `db` (the difference between these two will be explained later); and for each row of pixels, instead of writing [the bitplanes](../part1/tiles.md#encoding) as raw numbers, we will use a backtick (`` ` ``) followed by 8 characters.

Reviewer 3: ```suggestion
For each row of pixels, instead of writing [the bitplanes](../part1/tiles.md#encoding) directly, we will use a backtick (`` ` ``) followed by 8 characters.
```

Each character defines a single pixel, intuitively from left to right; it must be one of 0, 1, 2, and 3, representing the corresponding color index in [the palette](../part1/palettes.md).

::: tip

> 0, 1, 2, and 3 aren't the only options for writing graphics.
> You can use [`OPT g`](https://rgbds.gbdev.io/docs/v0.5.2/rgbasm.5/#Changing_options_while_assembling) to modify these characters to your liking.

Reviewer 3: ```suggestion
If the character selection isn't to your liking, you can use [RGBASM's `-g` option](https://rgbds.gbdev.io/docs/v0.5.2/rgbasm.1#g) or [`OPT g`](https://rgbds.gbdev.io/docs/v0.5.2/rgbasm.5/#Changing_options_while_assembling) to pick others.
For example, `rgbasm -g '.xXO' (...)` or `OPT g.xXO` would swap the four characters to `.`, `x`, `X`, and `O` respectively.
```

:::

For example:

```rgbasm
	dw `01230123 ; This is equivalent to `db $55,$33`
```

> 

Reviewer 3: ```suggestion

You may have noticed that we are using `dw` instead of `db`; the difference between these two will be explained later.
```

We already have tiles made for this project, so you can copy [this premade file](../../unbricked/getting-started/tileset.asm), and paste it at the end of your code.

Then copy the tilemap from [this file](https://github.com/ISSOtm/gb-asm-tutorial-part2/raw/main/tilemap.asm), and paste it after the `TilesEnd` label.

You can build the ROM now, by running the following commands in your terminal:

```console
$ rgbasm -L -o main.o main.asm
$ rgblink -o unbricked.gb main.o
$ rgbfix -v -p 0xFF unbricked.gb
```

If you run this in your emulator, you should see the following:

![Screenshot of our game](../assets/part2/img/screenshot.png)

That white square seems to be missing!
You may have noticed this comment earlier, somewhere in the tile data:

```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/getting-started/main.asm:custom_logo}}
{{#include ../../unbricked/getting-started/main.asm:custom_logo}}
```

The logo tiles were left intentionally blank so that you can choose your own.
You can use one of the following pre-made logos, or try coming up with your own!

> ## RGBDS Logo
> 
> ![The RGBDS Logo](../assets/part2/rgbds.png)
> 
> [Source](../../unbricked/getting-started/rgbds.asm)
> 
> ## Duck
> 
> ![A pixel-art duck](../assets/part2/duck.png)
> 
> [Source](../../unbricked/getting-started/duck.asm)
> 
> ## Tail
> 
> ![A silhouette of a tail](../assets/part2/tail.png)
> 
> [Source](../../unbricked/getting-started/tail.asm)

Reviewer 3: ```suggestion
- **RGBDS Logo**

  ![The RGBDS Logo](../assets/part2/rgbds.png)

  [Source](../../unbricked/getting-started/rgbds.asm)

- **Duck**

  ![A pixel-art duck](../assets/part2/duck.png)

  [Source](../../unbricked/getting-started/duck.asm)

- **Tail**

  ![A silhouette of a tail](../assets/part2/tail.png)

  [Source](../../unbricked/getting-started/tail.asm)
```
